### ISSUE TITLE
"idx" or "i" ? at  DPRDistributedSamplerWithValidation

### ISSUE THREAD

**Issue by xiahaoyun** [Zhylkaaa/LightningDPR#3]
```python
    def _is_valid(self, indexes):
        def concat_inputs(positives):
            return [((p['title'] + ' ') if p['title'] else '') + p['text'] for p in positives]

        current_contexts = set()

        for idx, i in enumerate(indexes):
            if (idx + 1) % self.disjoint_window_size == 0:
                current_contexts.clear()

            positives = concat_inputs(self.dataset[i]['positive_ctxs'])
            if len(current_contexts.intersection(set(positives))) > 0:
                return False
            else:
                current_contexts.update(set(positives))
        return True
```

https://github.com/Zhylkaaa/LightningDPR/blob/eac10bf82aeecf96d7b46e0e23ed12ee54145fe9/data_module.py#L91

I think this is dividing list `indexes` into consecutive, disjoint windows,size of every window is `disjoint_window_size `.
The purpose of the `_is_valid` is to make sure that there are no duplicate positive samples in each window.

For example, a list of length 12.  According to the subscript 0-3,4-7,8-12, they were divided into three groups, and no duplicate positive samples were allowed in each group.    

So I think it should be doing `current_contexts.clear()` when `i % self.disjoint_window_size == 0`.

What do you think, or maybe I misunderstood.

**Comment by Zhylkaaa** [comment]
Hello @xiahaoyun, thank you for your attention.
First, your understanding is correct `_is_valid` is making "sure" that there is no same positive contexts for `effective batch` (disjoint_window_size is set to `efective_batch_size` aka. `num_gpus * batch_size_per_gpu` later in code, referring to your example effective batch is 12 and batch_per_gpu is 4).

I think `idx` is correct because `i` is index of example in shuffled list and `idx` is `consecutive number` of this example in shuffled list, so on 13'th iteration `i` will be random index of example and `idx` will be 12. (while I was writing this I noted that `+1` is incorrect, and will fix it in nearest future)

Note: "sure" is in quotes because this algorithm is heuristic to reduce complexity and runtime (I think there are cases where this problem becomes NP-hard)

Thanks for pointing this out! 

(if you want to check the differences between this implementation and original DistributedSampler check `torch.utils.data.distributed.DistributedSampler` sources either on GitHub or in your code editor)

**Comment by Zhylkaaa** [comment]
fixed error, fill free to reopen this if I miss understood something

**Comment by xiahaoyun** [comment]
Thank you for your reply. 
I realized that I was confusing “idx” with “i”, because `enumerate(indexes)` was iterated over.
